### PR TITLE
Sim hardware

### PR DESCRIPTION
**Pull request by shaiTheKimhi**

### CODE REVIEW

File: Modules/Accelerometer/accelerometer_factory.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+if config.ACCELEROMETER_CONNECTED:
+    from . import mpu9250
+else:
+    from . import fake_accelerometer
+
+#this function returns the mpu9250 hardware module object if the hardware is connected.
+#otherwise returns and object that simulates hardware
+def initialize():
+    if(config.ACCELEROMETER_CONNECTED == True):
```
Comment (yotam180) [inline Modules/Accelerometer/accelerometer_factory.py]: No, you do not do it like that.

It's better to make two hidden functions, for example, `_init_mpu` and `_init_fake`. Then just assign `initialize = _init_mpu if config.ACCELEROMETER_CONNECTED else _init_fake`.

File: Modules/Accelerometer/fake_accelerometer.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class fake_accelerometer:
```
Comment (yotam180) [inline Modules/Accelerometer/fake_accelerometer.py:1]: Class documentation, please.

File: Modules/Accelerometer/mpu9250.py [hunk 3]
```diff
@@ -122,14 +122,14 @@ def read_xyz(self, address, register, lsb):
 		Reads x, y, and z axes at once and turns them into a tuple.
 		"""
 		# data is MSB, LSB, MSB, LSB ...
-                # For some reason when we read 6 bytes we get (0,0,0,0,0,0)
-                # So we changed it to 7 bytes like in an arduino library we found
-                # And not it works, go figure.
-                data = self.bus.read_i2c_block_data(address, register, 7)
+		# For some reason when we read 6 bytes we get (0,0,0,0,0,0)
```
Comment (yotam180) [inline Modules/Accelerometer/mpu9250.py:125]: You converted spaces to tabs on purpose??? Why not the other way around? :cry:

File: Modules/Field/circular_field.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import math
+
+from . import config
+
+def circular_field(angle:float):
```
Comment (yotam180) [inline Modules/Field/circular_field.py:5]: Definitely this is not UML oriented. You can ignore this comment for now, but we will have to restructure the algorithms & math system structure.

File: Modules/Gyro/fake_gyro.py [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class fake_gyro:
```
Comment (yotam180) [inline Modules/Gyro/fake_gyro.py:1]: Class documentation.

File: Modules/Gyro/gyro_factory.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+if config.GYRO_CONNECTED:
+    from . import mpu9250
+else:
+    from . import fake_gyro
+
+#this function returns the mpu9250 hardware module object if the hardware is connected.
+#otherwise returns and object that simulates hardware 
+def initialize():
+    if(config.GYRO_CONNECTED == True):
```
Comment (yotam180) [inline Modules/Gyro/gyro_factory.py]: Same comment as up there. The `GYRO_CONNECTED` check should be executed once as a preparation check and not in run-time.

File: Modules/Magnetometer/fake_magnetometer.py [hunk 7]
```diff
@@ -0,0 +1,8 @@
+from . import config
+
+class fake_magnetometer:
```
Comment (yotam180) [inline Modules/Magnetometer/fake_magnetometer.py]: Class documentation.

File: Modules/Magnetometer/magnetometer_factory.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    from . import mpu9250
+else:
+    from . import fake_magnetometer
+
+
+
+#this function returns the mpu9250 hardware module object if the hardware is connected.
+#otherwise returns and object that simulates hardware
+def initialize():
+    if(config.MAGNETOMETER_CONNECTED == True):
```
Comment (yotam180) [inline Modules/Magnetometer/magnetometer_factory.py]: Same note as above.

File: Tests/gyro.py [hunk 9]
```diff
@@ -1,26 +1,17 @@
...
-        msg = satprot.OutMsg(1,0)
-        msg.add_float(x)
-        msg.add_float(y)
-        msg.add_float(z)
         print ("x:%8.3f    y:%8.3f    z:%8.3f" % (x, y, z))
-        con.send(msg)
-    except KeyboardInterrupt:
-        shouldRun = False
+except KeyboardInterrupt:
+    shouldRun = False
```
Comment (yotam180) [inline Tests/gyro.py:17]: Does this line matter? From all the red-green blah blah, if your `try` contains the `while` and not the other way around, hitting an exception inside the `while` stops it regardless what you execute on your `except` clause.

**Review by yotam180 — CHANGES_REQUESTED** [review]
1) Comments
2) I have read from Slava & Avi that we would need to reshape a bit the system structure (mainly class relations and interactions). I will dig a bit more into the design you made there then we will redesign the structure to match their requests and remarks.

File: Tests/gyro.py (hunk 9 above)
Comment (ron-popov) [reply Tests/gyro.py:17]: Slava suggested this, just for easier readability

Comment (yotam180) [reply Tests/gyro.py:17]: Alright. Even though it is meaningless?

Comment (ron-popov) [reply Tests/gyro.py:17]: correct

File: Modules/Accelerometer/mpu9250.py (hunk 3 above)
Comment (ron-popov) [reply Modules/Accelerometer/mpu9250.py:125]: The whole file used tabs

Comment (yotam180) [reply Modules/Accelerometer/mpu9250.py:125]: :+1:

File: Modules/Accelerometer/accelerometer_factory.py (hunk 1 above)
Comment (ron-popov) [reply Modules/Accelerometer/accelerometer_factory.py]: The problem is i can't write these function before the import statements and i can't assign ```initialize``` to any of the methods before creating the methods.  
I wrote some code in 2c04131 but i don't think it's optimal.

Comment (yotam180) [reply Modules/Accelerometer/accelerometer_factory.py]: You're right, there are better ways to write it. However, it's still better than checking every time. You can leave it like that.

**Comment by ron-popov** [comment]
@yotam180 Can you review again ?

File: Modules/Accelerometer/fake_accelerometer.py (hunk 2 above)
Comment (yotam180) [reply Modules/Accelerometer/fake_accelerometer.py:1]: Just class documentation, please?

File: Modules/Gyro/fake_gyro.py (hunk 5 above)
Comment (yotam180) [reply Modules/Gyro/fake_gyro.py:1]: Also here por favor :slightly_smiling_face:

**Review by yotam180 — COMMENTED** [review]
Can you just add class documentations as I asked in the last review? Apart from that everything seems good.

Thanks.

File: Modules/Accelerometer/fake_accelerometer.py [hunk 10]
```diff
@@ -0,0 +1,6 @@
...
+    """Simulates an accelerometer sensor"""
```
Comment (ron-popov) [inline Modules/Accelerometer/fake_accelerometer.py:2]: Class Documentation @yotam180

File: Modules/Gyro/fake_gyro.py [hunk 11]
```diff
@@ -0,0 +1,6 @@
...
+    """Simulates a gyro sensor"""
```
Comment (ron-popov) [inline Modules/Gyro/fake_gyro.py:2]: Class documentation @yotam180

**Review by yotam180 — APPROVED** [review]
Great, seems ready to be merged 🙂

**Pull request merged**